File: DotNet/Messaging/IHydraService.cs
using System.Threading.Tasks;
using Shastra.Hydra.Messaging.Attachments;
using Shastra.Hydra.Messaging.Listeners;
using Shastra.Hydra.Messaging.MessageFetchers;
using Shastra.Hydra.Messaging.MessageIds;

namespace Shastra.Hydra.Messaging
{
    public interface IHydraService
    {
        IListener<TMessage> GetListener<TMessage>(IMessageFetcher<TMessage> messageFetcher, IMessageId startId = null, ListenerOptions listenerOptions = null) where TMessage : TransportMessage;
        Task<IMessageId> SendAsync<TMessage>(TMessage message) where TMessage : TransportMessage;
        Task<AttachmentContent> GetAttachmentAsync(Attachment attachment);
        string ServerName { get; }
    }
}

File: Java/hydra/src/main/java/uk/co/shastra/hydra/messaging/IHydraService.java
package uk.co.shastra.hydra.messaging;

import uk.co.shastra.hydra.messaging.listeners.IListener;
import uk.co.shastra.hydra.messaging.listeners.ListenerOptions;
import uk.co.shastra.hydra.messaging.messagefetchers.MessageFetcher;
import uk.co.shastra.hydra.messaging.messageids.MessageId;

public interface IHydraService {
	<TMessage extends TransportMessage> IListener<TMessage> getListener(MessageFetcher<TMessage> messageFetcher);
	<TMessage extends TransportMessage> IListener<TMessage> getListener(MessageFetcher<TMessage> messageFetcher, MessageId startId);
	<TMessage extends TransportMessage> IListener<TMessage> getListener(MessageFetcher<TMessage> messageFetcher, MessageId startId, ListenerOptions listenerOptions);
	<TMessage extends TransportMessage> MessageId send(TMessage message) throws Exception;
    String getServerName();
}
